library/alloc/src/rc.rs

```diff
@@ -3360,6 +3382,68 @@ fn data_offset_align(align: usize) -> usize {
     layout.size() + layout.padding_needed_for(align)
 }
 
+/// A unique owning pointer to a [`RcBox`] **that does not imply the contents are initialized,**
+/// but will deallocate it (without dropping the value) when dropped.
+///
+/// This is a helper for [`Rc::make_mut()`] to ensure correct cleanup on panic.
+#[cfg(not(no_global_oom_handling))]
+struct RcUninit<T: ?Sized, A: Allocator> {
```

Comment:
Oof, this is rough. I gather this only exists because `MaybeUninit` does not support ?Sized &mdash; otherwise this type wouldn't be needed and the code above would be much more straightforward with `Rc::new_uninit_in` + `clone_into_uninit` + `Rc::assume_init`. I do wish `MaybeUninit` worked with ?Sized some day.

Nicely done.

Reply:
> this only exists because `MaybeUninit` does not support `?Sized`.

Yes, that's precisely correct. It's also why `CloneToUnsized` can't take a `dst: &mut MaybeUninit<T>` (which otherwise would line up perfectly). At least `RcUninit` is only an implementation detail.